Added method to make creating faces on models less annoying

ArcAnc opened the pull request:
Added the ability to dynamically create all faces for the model, excluding sides that are unnecessary for the model

ArcAnc committed:
Update ModelBuilder.java

Added the ability to dynamically create all faces for the model, excluding sides that are unnecessary for the model

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/neoforged/NeoForge?pullRequest=1212) <br/>All committers have signed the CLA.

ArcAnc committed:
Update ModelBuilder.java

shartte commented:
@ArcAnc Your code does not pass the automated checks, please ensure it does.

And please give your PR a descriptive title.

HenryLoenwind commented on src/main/java/net/neoforged/neoforge/client/model/generators/ModelBuilder.java:
```diff
@@ -480,6 +480,24 @@ public ElementBuilder allFaces(BiConsumer<Direction, FaceBuilder> action) {
+         * @return this builder
+         * @throws NullPointerException if {@code action} is {@code null}
+         */
+        public ElementBuilder allFacesExcept(BiConsumer<Direction, FaceBuilder> action, Direction... exceptions) {
```
Why not `Set<Direction> exceptions`? You're putting it into a Collection anyway. And it's not that much more to type for the caller `allFacesExcept(..., EnumSet.of(Direction.NORTH, Direction.SOUTH));`. It also means that the caller can keep their set in a static field and avoid having an Array and a List object created on every call. EnumSets also have a much faster contains() lookup than Lists.

ArcAnc replied:
That's a good idea, I'll change it. I didn't think about it from the start

HenryLoenwind commented on src/main/java/net/neoforged/neoforge/client/model/generators/ModelBuilder.java:
```diff
@@ -480,6 +480,24 @@ public ElementBuilder allFaces(BiConsumer<Direction, FaceBuilder> action) {
+
+         /**
+         * Modify all <em>possible</em> faces except provided in second argument dynamically using a function, creating new
+         * faces as necessary.
```
"Modify all faces, except those provided in the second argument, dynamically using a function, creating new faces as necessary."

The description of allFaces() is already a bit wonky; your additions didn't make it better. That "all <em>possible</em>" should just be "<em>all</em>".  There are no "impossible faces"---the Direction enum is finite, all is all.

ArcAnc replied:
You are right, but sometimes may be situations, when some faces is not needed for model part. Like here:

```java
element().
     from(7, 6, 7).
     to(9, 7, 9).
     face(Direction.NORTH).end().
     face(Direction.SOUTH).end().
     face(Direction.WEST).end().
     face(Direction.EAST).end().
     face(Direction.UP).end().
     faces((face, builder) ->
     {
           builder.texture("#button");
           if (face != UP)
                builder.uvs(0, 8, 8, 12);
           else
                builder.uvs(0, 0, 8, 8);
      }).
      end()
```

In this situation, `Direction.DOWN` is useless for me, so to make the model a bit lighter and write less code with the new method, it will be easier to write:

```java
element().
     from(7, 6, 7).
     to(9, 7, 9).
     allFacesExcept((face, builder) ->
     {
           builder.texture("#button");
           if (face != UP)
                builder.uvs(0, 8, 8, 12);
           else
                builder.uvs(0, 0, 8, 8);
      }, EnumSet.of(Direction.DOWN)).
      end()
```


It might make sense to swap the list of exceptions and actions, but otherwise, everything is quite clear.

HenryLoenwind replied:
I was only suggesting a better javadoc text and ranting about the text of the existing allFaces. The concept was fine.

HenryLoenwind commented:
Hint: `gradlew applyAllFormatting` and `gradlew spotlessCheck`

ArcAnc committed:
Reworked allFacesExcept method for faster executing

Removed useless collection creation. In those variant it must be faster and allow users save faces for another model parts

ArcAnc committed:
Merge branch '1.21.x' into patch-1

sciwhiz12 commented:
Please do not create commits, particularly to code files, through the GitHub web editor. You should have a local dev environment where you modify the files, test your changes as appropriate, run other tasks such as formatting checks, and so on.

If you have trouble setting up an environment, please head over to the Discord server and ask for help.

ArcAnc commented:
> Please do not create commits, particularly to code files, through the GitHub web editor. You should have a local dev environment where you modify the files, test your changes as appropriate, run other tasks such as formatting checks, and so on.
> 
> If you have trouble setting up an environment, please head over to the Discord server and ask for help.

OK, sorry. This commit is quite small, so I thought it would be faster and more convenient to write it in the web version.
